### PR TITLE
support create recurring event

### PR DESCRIPTION
**Pull request by leedrum**
- support creating the recurring event 

<details>
<summary>example SOAP XML</summary>
<br>

```xml
<?xml version="1.0"?>
<soap:Envelope xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/" xmlns:t="http://schemas.microsoft.com/exchange/services/2006/types" xmlns:m="http://schemas.microsoft.com/exchange/services/2006/messages">
  <soap:Header>
    <t:RequestServerVersion Version="Exchange2010"/>
  </soap:Header>
  <soap:Body>
    <GetItem xmlns="http://schemas.microsoft.com/exchange/services/2006/messages">
      <ItemShape>
        <t:BaseShape>AllProperties</t:BaseShape>
      </ItemShape>
      <ItemIds>
        <t:ItemId Id="AAMkAGMzMDRhYjM2LThjOTktNDBkOS05MGM3LWI3Y2VkMjQ3NWRhZABGAAAAAACiZPxnvkqMRJTZYpH6kTO9BwAI1ITdf1huSo9J/THxcqsfAAAAAAFWAAAI1ITdf1huSo9J/THxcqsfAAAQ7NdGAAA=" ChangeKey="DwAAABYAAAAI1ITdf1huSo9J/THxcqsfAAAQ7OeT"/>
      </ItemIds>
    </GetItem>
  </soap:Body>
</soap:Envelope>
```
</details>

<details>
<summary>Success Response</summary>
<br>

```xml
<?xml version="1.0" encoding="utf-8"?>
<s:Envelope xmlns:s="http://schemas.xmlsoap.org/soap/envelope/">
  <s:Header>
    <h:ServerVersionInfo xmlns:h="http://schemas.microsoft.com/exchange/services/2006/types" xmlns="http://schemas.microsoft.com/exchange/services/2006/types" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" MajorVersion="15" MinorVersion="1" MajorBuildNumber="2507" MinorBuildNumber="23" Version="V2017_07_11"/>
  </s:Header>
  <s:Body xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema">
    <m:GetItemResponse xmlns:m="http://schemas.microsoft.com/exchange/services/2006/messages" xmlns:t="http://schemas.microsoft.com/exchange/services/2006/types">
      <m:ResponseMessages>
        <m:GetItemResponseMessage ResponseClass="Success">
          <m:ResponseCode>NoError</m:ResponseCode>
          <m:Items>
            <t:CalendarItem>
              <t:ItemId Id="AAMkAGMzMDRhYjM2LThjOTktNDBkOS05MGM3LWI3Y2VkMjQ3NWRhZABGAAAAAACiZPxnvkqMRJTZYpH6kTO9BwAI1ITdf1huSo9J/THxcqsfAAAAAAFWAAAI1ITdf1huSo9J/THxcqsfAAAQ7NdGAAA=" ChangeKey="DwAAABYAAAAI1ITdf1huSo9J/THxcqsfAAAQ7OeT"/>
              <t:ParentFolderId Id="AQMkAGMzMDRhYjM2LThjOTkALTQwZDktOTBjNy1iN2NlZDI0NzVkYWQALgAAA6Jk/Ge+SoxElNlikfqRM70BAAjUhN1/WG5Kj0n9MfFyqx8AAAIBVgAAAA==" ChangeKey="AQAAAA=="/>
              <t:ItemClass>IPM.Appointment</t:ItemClass>
              <t:Subject>tesst repeat event</t:Subject>
              <t:Sensitivity>Normal</t:Sensitivity>
              <t:Body BodyType="HTML">&lt;!DOCTYPE html&gt;&#13;
      &lt;html&gt;&#13;
      &lt;head&gt;&#13;
      &lt;meta http-equiv="Content-Type" content="text/html; charset=utf-8"&gt;&#13;
      &lt;/head&gt;&#13;
      &lt;body&gt;&#13;
      tesst repeat event&lt;br&gt;&#13;
      &lt;/body&gt;&#13;
      &lt;/html&gt;&#13;
      </t:Body>
              <t:DateTimeReceived>2023-05-15T07:29:37Z</t:DateTimeReceived>
              <t:Size>5585</t:Size>
              <t:Importance>Normal</t:Importance>
              <t:IsSubmitted>false</t:IsSubmitted>
              <t:IsDraft>false</t:IsDraft>
              <t:IsFromMe>false</t:IsFromMe>
              <t:IsResend>false</t:IsResend>
              <t:IsUnmodified>false</t:IsUnmodified>
              <t:DateTimeSent>2023-05-15T07:29:37Z</t:DateTimeSent>
              <t:DateTimeCreated>2023-05-15T07:29:37Z</t:DateTimeCreated>
              <t:ResponseObjects>
                <t:ForwardItem/>
              </t:ResponseObjects>
              <t:ReminderDueBy>2023-05-17T00:00:00Z</t:ReminderDueBy>
              <t:ReminderIsSet>true</t:ReminderIsSet>
              <t:ReminderMinutesBeforeStart>15</t:ReminderMinutesBeforeStart>
              <t:DisplayCc/>
              <t:DisplayTo/>
              <t:HasAttachments>false</t:HasAttachments>
              <t:Culture>de-DE</t:Culture>
              <t:EffectiveRights>
                <t:CreateAssociated>false</t:CreateAssociated>
                <t:CreateContents>false</t:CreateContents>
                <t:CreateHierarchy>false</t:CreateHierarchy>
                <t:Delete>true</t:Delete>
                <t:Modify>true</t:Modify>
                <t:Read>true</t:Read>
              </t:EffectiveRights>
              <t:LastModifiedName>ChartDesk-Service</t:LastModifiedName>
              <t:LastModifiedTime>2023-05-15T07:29:37Z</t:LastModifiedTime>
              <t:IsAssociated>false</t:IsAssociated>
              <t:WebClientReadFormQueryString>https://mail.ibtbunker.com/owa?itemid=AAMkAGMzMDRhYjM2LThjOTktNDBkOS05MGM3LWI3Y2VkMjQ3NWRhZABGAAAAAACiZPxnvkqMRJTZYpH6kTO9BwAI1ITdf1huSo9J%2FTHxcqsfAAAAAAFWAAAI1ITdf1huSo9J%2FTHxcqsfAAAQ7NdGAAA%3D&amp;exvsurl=1&amp;path=/calendar/item</t:WebClientReadFormQueryString>
              <t:ConversationId Id="AAQkAGMzMDRhYjM2LThjOTktNDBkOS05MGM3LWI3Y2VkMjQ3NWRhZAAQAKR+zT9NAGJEjV78ZiLPbtk="/>
              <t:UID>040000008200E00074C5B7101A82E00800000000BC06F400FF86D9010000000000000000100000009A1C4108E74CAB49BC589494EB98FBC0</t:UID>
              <t:DateTimeStamp>2023-05-15T07:29:37Z</t:DateTimeStamp>
              <t:Start>2023-05-03T00:00:00Z</t:Start>
              <t:End>2023-05-04T00:00:00Z</t:End>
              <t:IsAllDayEvent>true</t:IsAllDayEvent>
              <t:LegacyFreeBusyStatus>Busy</t:LegacyFreeBusyStatus>
              <t:IsMeeting>false</t:IsMeeting>
              <t:IsCancelled>false</t:IsCancelled>
              <t:IsRecurring>false</t:IsRecurring>
              <t:MeetingRequestWasSent>false</t:MeetingRequestWasSent>
              <t:IsResponseRequested>true</t:IsResponseRequested>
              <t:CalendarItemType>RecurringMaster</t:CalendarItemType>
              <t:MyResponseType>Organizer</t:MyResponseType>
              <t:Organizer>
                <t:Mailbox>
                  <t:Name>ChartDesk-Service</t:Name>
                  <t:EmailAddress>ChartDesk-Service@ibtbunker.com</t:EmailAddress>
                  <t:RoutingType>SMTP</t:RoutingType>
                  <t:MailboxType>Mailbox</t:MailboxType>
                </t:Mailbox>
              </t:Organizer>
              <t:Duration>P1D</t:Duration>
              <t:TimeZone>(UTC+00:00) Monrovia, Reykjavik</t:TimeZone>
              <t:AppointmentSequenceNumber>0</t:AppointmentSequenceNumber>
              <t:AppointmentState>0</t:AppointmentState>
              <t:Recurrence>
                <t:WeeklyRecurrence>
                  <t:Interval>1</t:Interval>
                  <t:DaysOfWeek>Wednesday</t:DaysOfWeek>
                </t:WeeklyRecurrence>
                <t:EndDateRecurrence>
                  <t:StartDate>2023-05-03Z</t:StartDate>
                  <t:EndDate>2023-05-30Z</t:EndDate>
                </t:EndDateRecurrence>
              </t:Recurrence>
              <t:FirstOccurrence>
                <t:ItemId Id="AAMkAGMzMDRhYjM2LThjOTktNDBkOS05MGM3LWI3Y2VkMjQ3NWRhZAFRAAgI20tpVp+AAEYAAAAAomT8Z75KjESU2WKR+pEzvQcACNSE3X9YbkqPSf0x8XKrHwAAAAABVgAACNSE3X9YbkqPSf0x8XKrHwAAEOzXRgAAEA==" ChangeKey="DwAAABYAAAAI1ITdf1huSo9J/THxcqsfAAAQ7OeT"/>
                <t:Start>2023-05-03T00:00:00Z</t:Start>
                <t:End>2023-05-04T00:00:00Z</t:End>
                <t:OriginalStart>2023-05-03T00:00:00Z</t:OriginalStart>
              </t:FirstOccurrence>
              <t:LastOccurrence>
                <t:ItemId Id="AAMkAGMzMDRhYjM2LThjOTktNDBkOS05MGM3LWI3Y2VkMjQ3NWRhZAFRAAgI21vp0UxAAEYAAAAAomT8Z75KjESU2WKR+pEzvQcACNSE3X9YbkqPSf0x8XKrHwAAAAABVgAACNSE3X9YbkqPSf0x8XKrHwAAEOzXRgAAEA==" ChangeKey="DwAAABYAAAAI1ITdf1huSo9J/THxcqsfAAAQ7OeT"/>
                <t:Start>2023-05-24T00:00:00Z</t:Start>
                <t:End>2023-05-25T00:00:00Z</t:End>
                <t:OriginalStart>2023-05-24T00:00:00Z</t:OriginalStart>
              </t:LastOccurrence>
              <t:IsOnlineMeeting>false</t:IsOnlineMeeting>
            </t:CalendarItem>
          </m:Items>
        </m:GetItemResponseMessage>
      </m:ResponseMessages>
    </m:GetItemResponse>
  </s:Body>
</s:Envelope>
```
</details>

<details>
<summary>response error</summary>
<br>

```xml
<?xml version="1.0" encoding="utf-8"?>
<s:Envelope xmlns:s="http://schemas.xmlsoap.org/soap/envelope/">
  <s:Header>
    <h:ServerVersionInfo xmlns:h="http://schemas.microsoft.com/exchange/services/2006/types" xmlns="http://schemas.microsoft.com/exchange/services/2006/types" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" MajorVersion="15" MinorVersion="1" MajorBuildNumber="2507" MinorBuildNumber="23" Version="V2017_07_11"/>
  </s:Header>
  <s:Body xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema">
    <m:CreateItemResponse xmlns:m="http://schemas.microsoft.com/exchange/services/2006/messages" xmlns:t="http://schemas.microsoft.com/exchange/services/2006/types">
      <m:ResponseMessages>
        <m:CreateItemResponseMessage ResponseClass="Error">
          <m:MessageText>Ein Serienelement des Kalenderserienelements überschneidet sich mit einem anderen Serienelement des gleichen Kalenderelements., Die Dauer zwischen „startTime“ und „endTime“ (27.00:00:00) der Serie ist größer als die Mindestdauer zwischen zwei Serienelementen (21.00:00:00).</m:MessageText>
          <m:ResponseCode>ErrorOccurrenceTimeSpanTooBig</m:ResponseCode>
          <m:DescriptiveLinkKey>0</m:DescriptiveLinkKey>
          <m:Items/>
        </m:CreateItemResponseMessage>
      </m:ResponseMessages>
    </m:CreateItemResponse>
  </s:Body>
</s:Envelope>
```
</details>

### CODE REVIEW

File: lib/ews/soap/builders/ews_builder.rb [hunk 1]
```diff
@@ -870,10 +870,70 @@ def weekly_recurrence!(item)
...
+
+    def absolute_yearly_recurrence!(item)
+      nbuild[NS_EWS_TYPES].absoluteYearly {
+        item.each_pair { |k, v|
+          self.send("#{k}!", v)
+        }
+      }
+    end
+
+    def relative_yearly_recurrence!(item)
```
Comment (haanhduclinh) [inline lib/ews/soap/builders/ews_builder.rb:897]: ? I think function name should be
```ruby
def relative_monthly_recurrence
```

File: lib/ews/soap/builders/ews_builder.rb [hunk 2]
```diff
@@ -870,10 +870,70 @@ def weekly_recurrence!(item)
...
+    def relative_monthly_recurrence(item)
+      nbuild[NS_EWS_TYPES].RelativeMonthlyRecurrence {
+        item.each_pair { |k, v|
+          self.send("#{k}!", v)
+        }
+      }
+    end
+
+    def absolute_yearly_recurrence!(item)
+      nbuild[NS_EWS_TYPES].absoluteYearly {
```
Comment (haanhduclinh) [inline lib/ews/soap/builders/ews_builder.rb]: is that valid xml request? could you please give me url of document?

File: lib/ews/soap/builders/ews_builder.rb [hunk 3]
```diff
@@ -870,10 +870,70 @@ def weekly_recurrence!(item)
       }
     end
 
-    def interval!(num)
+    def absolute_monthly_recurrence!(item)
+      nbuild[NS_EWS_TYPES].AbsoluteMonthRecurrence {
```
Comment (haanhduclinh) [inline lib/ews/soap/builders/ews_builder.rb]: do you mean this field? [AbsoluteMonthlyRecurrence](https://learn.microsoft.com/en-us/exchange/client-developer/web-service-reference/absolutemonthlyrecurrence)

File: lib/ews/soap/builders/ews_builder.rb [hunk 4]
```diff
@@ -870,10 +870,70 @@ def weekly_recurrence!(item)
...
+
+    def day_of_month!(day_of_month) # example: 1..31
+      nbuild[NS_EWS_TYPES].DayOfMonth(day_of_month)
+    end
+
+    def day_of_week_index!(day) # ex: Monday, Tuesday, Wednesday, Thursday, Friday, Saturday, Sunday
+      nbuild[NS_EWS_TYPES].DayOfWeekIndex(day)
+    end
+
+    def month!(month) # ex: 1..12
```
Comment (haanhduclinh) [inline lib/ews/soap/builders/ews_builder.rb:925]: [Q] I saw the document. This filed only accept text [Month (Item Recurrence)](https://learn.microsoft.com/en-us/exchange/client-developer/web-service-reference/month-item-recurrence). can we use number in it?

File: lib/ews/soap/builders/ews_builder.rb [hunk 5]
```diff
@@ -870,10 +870,70 @@ def weekly_recurrence!(item)
...
+      nbuild[NS_EWS_TYPES].Month(day_of_month)
+    end
+
+    def interval!(num) # 1..99
       nbuild[NS_EWS_TYPES].Interval(num)
     end
 
+    def is_recurring!(boolean)
+      nbuild[NS_EWS_TYPES].IsRecurring(boolean.upcase)
```
Comment (haanhduclinh) [inline lib/ews/soap/builders/ews_builder.rb]: [IMO]
```
boolean.to_s.upcase
```

**Review by haanhduclinh — CHANGES_REQUESTED** [review]

File: lib/ews/soap/builders/ews_builder.rb (hunk 1 above)
Comment (leedrum) [reply lib/ews/soap/builders/ews_builder.rb:897]: it's should be `relative_yearly_recurrence `

File: lib/ews/soap/builders/ews_builder.rb (hunk 4 above)
Comment (leedrum) [reply lib/ews/soap/builders/ews_builder.rb:925]: fixed. using `Date::MONTHNAMES[month.to_i]`

**Review by haanhduclinh — APPROVED** [review]
LGTM!

**Pull request merged**